crates/noirc_evaluator/src/ssa/context.rs

```diff
@@ -1020,16 +1071,122 @@ impl SsaContext {
+        }
+
+        let (len, elem_type) = function::get_result_type(opcode);
+        let array_set = if len > 1 {
+            let array = self.new_array(&format!("{}_result", opcode), elem_type, len, None).1;
+            self.push_array_set(vec![(array, 0)])
+        } else {
+            self.push_array_set(vec![])
+        };
+
+        let index = self.nodes.insert_with(|index| {
+            let node_id = NodeId(index);
+            let typ = ObjectType::Function(array_set);
+            NodeObj::Function(FunctionKind::Builtin(opcode), node_id, typ)
+        });
+        self.opcode_ids.insert(opcode, NodeId(index));
+        NodeId(index)
+    }
+
+    pub fn get_builtin_opcode(&self, node_id: NodeId) -> Option<OPCODE> {
+        match &self[node_id] {
+            NodeObj::Function(FunctionKind::Builtin(opcode), ..) => Some(*opcode),
+            _ => None,
+        }
+    }
+
+    pub fn push_array_set(&mut self, arrays: Vec<(ArrayId, u32)>) -> ArrayIdSet {
+        let next_id = self.function_returned_arrays.len();
+        self.function_returned_arrays.push(arrays);
+        ArrayIdSet(next_id as u32)
+    }
+
+    pub fn union_array_sets(&mut self, first: ArrayIdSet, second: ArrayIdSet) -> ArrayIdSet {
+        let mut union_set = self.function_returned_arrays[first.0 as usize].clone();
+        union_set.extend_from_slice(&self.function_returned_arrays[second.0 as usize]);
+        union_set.sort();
+        union_set.dedup();
+        self.push_array_set(union_set)
+    }
+
+    pub fn get_returned_arrays(&self, func: NodeId) -> Vec<(ArrayId, u32)> {
+        match self[func].get_type() {
+            ObjectType::Function(arrays) => {
+                self.function_returned_arrays[arrays.0 as usize].clone()
+            }
+            other => unreachable!("get_returned_arrays: Expected function type, found {:?}", other),
+        }
+    }
+
+    pub fn convert_type(&mut self, t: &noirc_frontend::monomorphisation::ast::Type) -> ObjectType {
+        use noirc_frontend::monomorphisation::ast::Type;
+        use noirc_frontend::Signedness;
+        match t {
+            Type::Bool => ObjectType::Boolean,
+            Type::Field => ObjectType::NativeField,
+            Type::Integer(sign, bit_size) => {
+                assert!(
+                    *bit_size < super::integer::short_integer_max_bit_size(),
+                    "long integers are not yet supported"
+                );
+                match sign {
+                    Signedness::Signed => ObjectType::Signed(*bit_size),
+                    Signedness::Unsigned => ObjectType::Unsigned(*bit_size),
+                }
+            }
+            // TODO: We should not track arrays through ObjectTypes
+            Type::Array(..) => ObjectType::Pointer(ArrayId::dummy()),
+            Type::Unit => ObjectType::NotAnObject,
+            Type::Function(..) => {
+                // TODO: This is incorrect, we cannot track arrays through function types in this case
```

Comment:
It is incorrect because ArrayIds should refer to array values and thus can never be guessed by converting from a type which could be the type of many possible values. This is the same reasoning as the TODO on line 1178.

Reply:
Thanks for adding the info to the issue. Should we not panic here then if we cannot convert an array `Type` into an `ObjectType`. Or is the usage of dummy and an empty array vec ultimately giving us the error we need?

Reply:
I think the function case pushing an empty array may be needed for now until there is a refactoring. I've already removed all instances of us converting an array type directly into an ObjectType (as in all cases we just wanted the array element type previously). So these can safely panic I believe.